packages/core/src/browser/tree/tree-decorator.ts

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from 'inversify';
+import { Event, Emitter } from '../../common/event';
+
+/**
+ * CSS styles for the tree decorators.
+ */
+export namespace DecoratorStyles {
+    export const CAPTION_PREFIX = 'theia-caption-prefix';
+    export const CAPTION_SUFFIX = 'theia-caption-suffix';
+    export const ICON_WRAPPER = 'theia-icon-wrapper';
+    export const DECORATOR_SIZE = 'theia-decorator-size';
+    export const TOP_RIGHT = 'theia-top-right';
+    export const BOTTOM_RIGHT = 'theia-bottom-right';
+    export const BOTTOM_LEFT = 'theia-bottom-left';
+    export const TOP_LEFT = 'theia-top-left';
+}
+
+/**
+ * Tree decorator service that can change the outlook and the style of the tree items within a widget.
+ */
+export const TreeDecorator = Symbol('TreeDecorator');
+export interface TreeDecorator {
+
+    /**
+     * The unique identifier of the decorator. Ought to be unique in the application.
+     */
+    readonly id: string;
+
+    /**
+     * Fired when the decoration has calculated all the decoration data. Keys are the identifiers of the tree items.
+     */
+    onDidChangeDecorations: Event<Map<string, DecorationData>>;
```

Comment:
How should `TreeDecorator` not about node id? Is there a way that it gets it from a node in that way changing the computation of node id does not break all decorators. I think decorators computation should depend on the tree state. It could be enen that decorators should not be computed at all if a node is not visible to the user.

Reply:
> It could be enen that decorators should not be computed at all if a node is not visible to the user.

It is not about overcomplicating the things, but I might want to add the error marker to a folder if any of the containing files has an error even if the folder is not in the expanded state.

Reply:
I meant you don't need to compute decorators for children then only for folders.

Reply:
The decoration data is not computed by the tree or the tree model, it is pushed by the decorator once it has been calculated, then we re-render the tree.